feat: add accessor array support to `blas/base/gnrm2`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: na
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: passed ---

Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/26

## Description

> What is the purpose of this pull request?

This pull request:

-   adds accessor array support to `blas/base/gnrm2`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/26

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add accessor array support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

headlessNode committed:
fix: add missing docs and refactor

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/README.md:
````diff
@@ -87,7 +87,7 @@ var z = gnrm2( 4, x1, 2 );
 ```
 
-If either `N` or `stride` is less than or equal to `0`, the function returns `0`.
+If `N` parameter is less than or equal to `0`, the function returns `0`.
````
```suggestion
If `N` is less than or equal to `0`, the function returns `0`.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/lib/accessors.js:
```diff
@@ -0,0 +1,89 @@
+/**
+* Computes the L2-norm of a vector.
+*
+* @param {PositiveInteger} N - number of indexed elements
```
```suggestion
* @private
* @param {PositiveInteger} N - number of indexed elements
```

kgryte committed:
docs: fix missing private annotation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/lib/accessors.js:
```diff
@@ -0,0 +1,90 @@
+		return abs( get( buf, ix ) );
+	}
+	if ( stride === 0 ) {
+		return abs( N * get( buf, ix ) );
```
@headlessNode This isn't correct. Think about the quantity being computed. I suggest writing out with pencil and paper.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/lib/ndarray.js:
```diff
@@ -37,28 +39,35 @@ var pow = require( '@stdlib/math/base/special/pow' );
+		return abs( x[ ix ] );
+	}
+	if ( stride === 0 ) {
+		return abs( N * x[ ix ] );
```
@headlessNode This isn't correct. Think about the quantity being computed. I suggest writing out with pencil and paper.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.main.js:
```diff
@@ -87,32 +104,102 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
 
 	z = gnrm2( x.length, x, 0 );
-	t.strictEqual( z, 0.0, 'returns expected value' );
+	t.strictEqual( z, 5.0, 'returns expected value' );
```
This isn't correct.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.main.js:
```diff
@@ -87,32 +104,102 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+	x = [ 1.0, -2.0, -4.0, 5.0, 3.0 ];
+
+	z = gnrm2( x.length, toAccessorArray( x ), 0 );
+	t.strictEqual( z, 5.0, 'returns expected value' );
```
This isn't correct.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.main.js:
```diff
@@ -87,32 +104,102 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector repeated `N` times', function test( t ) {
```
This description will eventually need updating. Same comment for L188.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.ndarray.js:
```diff
@@ -109,15 +124,79 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector repeated `N` times', function test( t ) {
```
Same comments as test.js.

headlessNode committed:
fix: bug when stride 0

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

headlessNode committed:
fix: test case description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/lib/accessors.js:
```diff
@@ -0,0 +1,90 @@
+		return abs( get( buf, ix ) );
+	}
+	if ( stride === 0 ) {
+		return sqrt( N * pow( abs( get( buf, ix ), 2 ) ) );
```
This is still not correct. Why do you need to compute `sqrt(pow(x,2))`?

kgryte replied:
As this is now, you are at risk of overflow. Something that this function is explicitly intended to guard against.

kgryte replied:
Hint: (xy)^z = x^z y^z

headlessNode replied:
You mean like this? `sqrt( N ) * abs( get( buf, ix ) )`

kgryte replied:
Precisely.

kgryte replied:
With that formulation, the result only overflows if, in fact, the result overflows, not because of an overflow of intermediate values.

headlessNode replied:
Yeah it took me a minute to understand😅 Thanks for the hint @kgryte

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.main.js:
```diff
@@ -87,32 +105,102 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector multiplied by `N` square root', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of a vector', function test( t ) {
```

Your definition is not correct. Better to avoid fitting a square peg into a round hole on this one.

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.main.js:
```diff
@@ -87,32 +105,102 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector multiplied by `N` square root (accessors)', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of a vector (accessors)', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.ndarray.js:
```diff
@@ -109,15 +125,79 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector multiplied by `N` square root', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of a vector', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/gnrm2/test/test.ndarray.js:
```diff
@@ -109,15 +125,79 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
+	t.end();
+});
+
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of the first element of the vector multiplied by `N` square root (accessors)', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the L2-norm of a vector (accessors)', function test( t ) {
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
